File: 059_spiral_matrix_II.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] res = new int[n][n];
        int max = n * n, start = 1;
        int M = n, N = n;
        int row = 0, col = -1;
        while (start <= max) {
            // move right
            for (int i = 0; i < N; i++) {
                res[row][++col] = start++;
            }
            if (--M == 0) break;

            // move down
            for (int i = 0; i < M; i++) {
                res[++row][col] = start++;
            }
            if (--N == 0) break;

            // move left
            for (int i = 0; i < N; i++) {
                res[row][--col] = start++;
            }
            if (--M == 0) break;

            // move up
            for (int i = 0; i < M; i++) {
                res[--row][col] = start++;
            }
            if (--N == 0) break;
        }
        return res;
    }
}

File: python/059_spiral_matrix_II.py
class Solution:
    # Time Complexity: O(n ^ 2)
    # Space Complexity: O(n ^ 2)
    def generateMatrix(self, n: int) -> List[List[int]]:
        res = [[0 for _ in range(n)] for _ in range(n)]
        M, N = n, n
        row, col = 0, -1
        num_eles = n ** 2
        curr_num = 1
        while curr_num <= num_eles:
            # move right
            for i in range(0, N):
                col += 1
                res[row][col] = curr_num
                curr_num += 1
            M -= 1
            if M == 0: break

            # move down
            for i in range(0, M):
                row += 1
                res[row][col] = curr_num
                curr_num += 1
            N -= 1
            if N == 0: break

            # move left
            for i in range(0, N):
                col -= 1
                res[row][col] = curr_num
                curr_num += 1
            M -= 1
            if M == 0: break

            # move up
            for i in range(0, M):
                row -= 1
                res[row][col] = curr_num
                curr_num += 1
            N -= 1
            if N == 0: break
        return res
